presto-parquet/src/main/java/io/prestosql/parquet/reader/PrimitiveColumnReader.java

```diff
@@ -79,47 +94,89 @@ protected boolean isValueNull()
+                }
+                if (descriptor.getPrimitiveType().getPrimitiveTypeName() == PrimitiveTypeName.INT64) {
+                    return new LongColumnReader(descriptor, prestoType);
                 }
-                return createDecimalColumnReader(descriptor).orElse(new LongColumnReader(descriptor));
-            case INT96:
-                return new TimestampColumnReader(descriptor);
-            case FLOAT:
-                return new FloatColumnReader(descriptor);
-            case DOUBLE:
-                return new DoubleColumnReader(descriptor);
-            case BINARY:
-                return createDecimalColumnReader(descriptor).orElse(new BinaryColumnReader(descriptor));
-            case FIXED_LEN_BYTE_ARRAY:
-                return createDecimalColumnReader(descriptor)
-                        .orElseThrow(() -> new PrestoException(NOT_SUPPORTED, " type FIXED_LEN_BYTE_ARRAY supported as DECIMAL; got " + descriptor.getPrimitiveType().getOriginalType()));
-            default:
-                throw new PrestoException(NOT_SUPPORTED, "Unsupported parquet type: " + descriptor.getPrimitiveType().getPrimitiveTypeName());
+            }
+        }
+
+        if (prestoType instanceof DecimalType) {
+            if (parquetDecimalType.isPresent() && ((DecimalType) prestoType).getScale() == parquetDecimalType.get().getScale()) {
+                // TODO allow rescaling
+                return decimalColumnReader.orElseThrow(() -> new VerifyException("decimalColumnReader not present when parquetDecimalType is present")).get();
+            }
+        }
+
+        if (prestoType == RealType.REAL && descriptor.getPrimitiveType().getPrimitiveTypeName() == PrimitiveTypeName.FLOAT) {
+            return new FloatColumnReader(descriptor, prestoType);
+        }
+
+        if (prestoType == DoubleType.DOUBLE && descriptor.getPrimitiveType().getPrimitiveTypeName() == PrimitiveTypeName.DOUBLE) {
+            return new DoubleColumnReader(descriptor, prestoType);
         }
+
+        if (prestoType == DateType.DATE && descriptor.getPrimitiveType().getPrimitiveTypeName() == PrimitiveTypeName.INT32) {
+            // TODO verify no truncation occurs
+            return new IntColumnReader(descriptor, prestoType);
+        }
+
+        if (prestoType == TimestampType.TIMESTAMP && descriptor.getPrimitiveType().getPrimitiveTypeName() == PrimitiveTypeName.INT96) {
+            return new TimestampColumnReader(descriptor, prestoType);
+        }
+
+        if (prestoType == TimestampType.TIMESTAMP && descriptor.getPrimitiveType().getPrimitiveTypeName() == PrimitiveTypeName.INT64 && descriptor.getPrimitiveType().getOriginalType() == OriginalType.TIMESTAMP_MICROS) {
+            return new TimestampMicrosColumnReader(descriptor, prestoType);
+        }
+
+        if (prestoType instanceof VarcharType && descriptor.getPrimitiveType().getPrimitiveTypeName() == PrimitiveTypeName.BINARY) {
```

Comment:
can we collapse this 3 if clauses?

Reply:
We could. Except they shouldn't really be handled the same way. This is why i kept them separate.